lib/okta_oidc.go

```diff
@@ -0,0 +1,249 @@
+		)
+		if err != nil {
+			log.Errorf("Failed to exchange OIDC token: %s", err.Error())
+			return
+		}
+		rawIDToken, ok := oauth2Token.Extra("id_token").(string)
+		if !ok {
+			log.Error("No id_token field in oauth2 token")
+			return
+		}
+		_, err = verifier.Verify(ctx, rawIDToken)
+		if err != nil {
+			log.Errorf("Failed to verify OIDC token: %s", err.Error())
+			return
+		}
+
+		log.Debug("Verified OIDC token. Using the ID token to assume an AWS role.")
+		sess := session.Must(session.NewSession())
+		svc := sts.New(sess)
+		params := &sts.AssumeRoleWithWebIdentityInput{
+			RoleArn:          aws.String(profileARN),
+			RoleSessionName:  aws.String(o.Username),
+			WebIdentityToken: aws.String(rawIDToken),
+			DurationSeconds:  aws.Int64(int64(duration.Seconds())),
+		}
+		assumeResp, err := svc.AssumeRoleWithWebIdentity(params)
+		if err != nil {
+			log.WithField("role", profileARN).Errorf("Error assuming role with OIDC: %s", err.Error())
+			return
+		}
+		credChan <- assumeResp.Credentials
+	})
+}
+
+func (o *OktaOIDCClient) startRedirectServer() *http.Server {
+	srv := &http.Server{Addr: RedirectAddr}
+	go func() {
+		log.Debugf("Listening for redirect on http://%s/", RedirectAddr)
+		if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			log.Debugf("Httpserver: ListenAndServe() error: %s", err)
+		}
+	}()
+	return srv
+}
+
+func (o *OktaOIDCClient) stopRedirectServer(srv *http.Server) {
+	log.Debug("Shutting server down")
+	if err := srv.Close(); err != nil {
+		// This should be safe to ignore in a CLI program, which shuts down
+		// shortly after anyway
+		log.Debug("Failed to shut down server. Ignoring")
+	}
+}
+
+func restoreWindowFocusAfter(f func() error) error {
+	if runtime.GOOS != "darwin" {
+		log.Debugf("Not runninc on macOS. Will not try to restore focus to terminal.")
+		f()
+		return nil
+	}
+
+	windowWithFocusRaw, err := exec.Command("osascript", "-e", "tell application \"System Events\""+
+		" to return name of first application process whose frontmost is true").Output()
+	if eErr, ok := err.(*exec.ExitError); ok {
+		log.Debugf(
+			"Getting current focused window failed. Will not try restoring focus. Error: %s",
+			string(eErr.Stderr),
+		)
+	} else if err != nil {
+		return err
+	}
+	windowWithFocus := strings.TrimSpace(string(windowWithFocusRaw))
+
+	// Don't return immedately, even if f returns an error. Try to restore
+	// focus first.
+	fErr := f()
+
+	if err != nil {
+		// Do not try to regain focus if finding the window to focus failed
+		return fErr
+	}
+
+	err = exec.Command("osascript", "-e",
+		fmt.Sprintf("tell application \"%s\" to activate", windowWithFocus)).Run()
+	if eErr, ok := err.(*exec.ExitError); ok {
+		log.Debugf("Refocusing failed. Ignoring. Error: %s", string(eErr.Stderr))
+	} else if err != nil {
+		return err
+	}
+	return fErr
+}
+
+func pkce() (string, string, error) {
+	codeVerifier, err := randomHex(30)
```

Comment:
Why 30?

Reply:
And why hex?

Based on [this doc](https://www.oauth.com/oauth2-servers/pkce/authorization-reque), the string can contain `[A-Za-z0-9._~-]`, and can be 43-128 characters long